Module type substitutions

Octachron opened the pull request:
This PR adds the possibility to use `with` substitutions on module types (both the destructive and non-destructive variants).

For instance:
```ocaml
module type ENDO = sig
  module type T
  module F: T -> T
end
module Endo(X: sig module type T end): ENDO with module type T = X.T = struct
  module type T = X.T
  module F(X:T) = X
end
```
This mirrors the preexisting substitutions for types and modules, even if the feature itself is probably at most moderately useful.

In term of design , the right hand side of the `with` constraint is limited to path expressions, in order to ensure that substitution preserves valid module types for first class modules. Thus
```ocaml
module type S = sig
  module type T
  val x: (module T)
end
with module type T := RHS
```
always results in a valid module type for all valid `RHS` module types.

Another constraint is that for non-abstract left-hand-side module type, substitution is only allowed if the two right-hand side module types are equivalent. Typically,
```ocaml
module type T = sig
  module type Inner = sig type a type b end
  module type Equi_inner = sig include Inner end
  module type No_c = sig include Inner type c end
  module Always_a: functor (X:Inner) -> sig type a = X.a end 
end
with module type Inner = RHS
```
is only valid if `RHS` ~  `sig type a type b end ` .

This constraints keeps the substitution well-behaved:
- `Equi_inner` is always equivalent to `Inner` .
- both `No_c` and `Always_a` are always well-typed.

Octachron committed:
Module type substitutions: core

Octachron committed:
module type substitutions: syntax

Octachron committed:
module type substitutions: documentation

Octachron committed:
Update Changes

gasche commented:
Three naive comments:

- It would make sense to allow a wider range of equality/substitution right-hand-sides, but fail if they are used on those module signatures where only paths make sense. This is also more implementation work and safety scares, so I am completely fine with the restricted scope of this PR as long as there is no clear demand.
- I wonder how those constructs map into @Drup's finer-grained module calculus.
- In the non-abstract case, is there a use for adding equalities on equivalent signatures? I think it makes sense to have the feature just to be consistent/unsurprising, but I still wonder if there is an actual use-case -- intuitively I would expect that, when `Foo` is not abstract, `with module type Foo = Bar` should be a no-op in the case where `Foo` and `Bar` are equivalent -- the non-equated module should work in the same settings as the equated one.

lpw25 commented:
> In the non-abstract case, is there a use for adding equalities on equivalent signatures?

One reason to allow this is that otherwise you risk getting errors when exposing the definition of an abstract module type -- something that should really always be safe to do. A warning that the equality is not needed would be reasonable, but I think an error would be a mistake.

lpw25 commented:
I'm very much in favour of having this. Could we get the corresponding destructive substitution in signatures as well?
```ocaml
module M : sig
  module type S := sig type t end
  module N : S
end
```
It would be good to keep those in sync, and I've had quite a few requests for it. To be really useful it needs to let you use it with non-path definitions, but you should be able to just error if a non-path definition appears in a first-class module type.

Possibly it should be in a separate PR, but aiming for the same OCaml release.

Octachron commented:
I can offer a small use case for the non-abstract case: when doing multiple substitutions, it may make sense to use an module type substitution to check an equality that is dependent on the earlier substitution. For instance,
```ocaml
module type S = sig
  type x
  module type T = sig type t = X of x end
end
module type U = S with type x = M.x and module type T = M.T
```
I did not think of (inline) destructive substitution in signatures. I will have a look at adding them.

Octachron committed:
module type S := ...: core

Octachron committed:
Subst: add_modtype_path function

Deep destructive substitution might require to substitute whole module type
paths rather than just identifiers.

Octachron committed:
Fix handling of deeply nested substitution

Octachron committed:
with module type = sig end

Allow non-path right-hand side for module type with constraints.

Octachron committed:
add syntax for local module type substitutions

Octachron committed:
testsuite: module type substitutions

Octachron committed:
manual: module type substitutions

Octachron committed:
update changes

Octachron commented:
I have added the support for local module type substitutions, and lifted the restriction on the right-hand side module types in both cases (uses of substituted away module types in packed module type raises an error).

lpw25 commented on parsing/parsetree.mli:
```diff
@@ -850,6 +852,10 @@ and with_constraint =
            the name of the type_declaration. *)
   | Pwith_module of Longident.t loc * Longident.t loc
         (* with module X.Y = Z *)
+  | Pwith_module_type of Longident.t loc * module_type
```
`Pwith_modtype` would be more consistent

Octachron replied:
I also updated the typedtree version.

lpw25 commented on parsing/parsetree.mli:
```diff
@@ -850,6 +852,10 @@ and with_constraint =
         (* with module X.Y = Z *)
+  | Pwith_module_type of Longident.t loc * module_type
+        (* with module type X.Y = Z *)
+  | Pwith_module_typesubst of Longident.t loc * module_type
```
`Pwith_modtypesubst` would be more consistent (I know it is ugly)

lpw25 commented on parsing/pprintast.ml:
```diff
@@ -1145,14 +1152,18 @@ and signature_item ctxt f x : unit =
         (item_attributes ctxt) incl.pincl_attributes
-  | Psig_modtype {pmtd_name=s; pmtd_type=md; pmtd_attributes=attrs} ->
+  | Psig_modtype {pmtd_name=s; pmtd_type=md; pmtd_attributes=attrs}
+  | Psig_modtypesubst {pmtd_name=s; pmtd_type=md; pmtd_attributes=attrs} as p ->
```
I'd prefer to split these cases and keep everything explicit.

lpw25 commented on testsuite/tests/typing-modules/module_type_substitution.ml:
```diff
@@ -0,0 +1,270 @@
+Line 1, characters 25-58:
+1 | module type fst_erased = fst with module type t := sig end
+                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: This `with' constraint t := sig end makes a packed module ill-typed.
```
I think something like "ill-formed" would be better than "ill-typed" here.

lpw25 commented on typing/typemod.ml:
```diff
@@ -373,9 +373,48 @@ let check_usage_of_path_of_substituted_item paths env signature ~loc ~lid =
+  (* The last item is the one that's removed. We don't need to check how
+        it's used since it's replaced by a more specific type/module. *)
+  assert (match last with Pident _ -> true | _ -> false);
+  let deep_path_check it = match rest with
```
Building these two closures just to immediately apply them is convoluted. Why not just bind the iterator as `iterator` instead of `super` at the top and then have two `let iterator =` to extend the iterator with the checks?

Octachron replied:
I liked to have explicit names, but it is probably better to avoid the intermediary closures.

lpw25 commented on typing/typemod.ml:
```diff
@@ -448,18 +487,20 @@ let params_are_constrained =
   loop
 ;;
 
-let merge_constraint initial_env remove_aliases loc sg constr =
```
I think we should avoid passing in `transl_modtype` here. One possibility would be to just move `merge_constraints` down into the mutually recursive block containing `transl_modtype`. However, my preferred approach would be to separate out the typing of the with-constraints with merging them into a signature. In other words, `merge_constraints` should operate on `Typedtree.with_constraint` instead of `Parsetree.with_constraint` and there should be a `transl_with_constraint` function that is mutually recursive with `transl_modtype`.

(I would actually ideally prefer it to use a separate type -- living in types.ml -- containing only the information required by `merge_constraints`, but I know that not everyone shares my distaste for re-examining translated parsetrees and I don't want to add too much additional work to this PR).

Octachron replied:
Operating with the typedtree of the right hand side works fine with the module and module types substitutions,  but not with the type substitutions: `Typedecl.transl_with_constraint` merges the original signature type with the parsetree type declaration on right hand side of the substitution.

I took an intermediary path with a `with_info` type that carries the information needed by `merge_constraints`.

lpw25 replied:
I actually think it would be worth spending the time to split `Typedecl.transl_with_constraint` into two pieces, one that translates the with constraint and another that merges the with constraint with a declaration. However, I think it is fine to leave that for a later PR.

lpw25 commented on typing/typemod.ml:
```diff
@@ -1010,6 +1094,30 @@ end = struct
+    should raise an error.
+   *)
+  let check_unpackable_modtypes ~loc ~env to_remove component =
+    if not (Ident.Set.is_empty to_remove.unpackable_modtypes) then
```
Please use `begin ... end` to make the extent of this `if` clear.

lpw25 commented on typing/typemod.ml:
```diff
@@ -1398,6 +1510,22 @@ and transl_signature env sg =
+            let info id tmty =
+              let mty = match tmty with
+                | Some tmty -> tmty.mty_type
+                | None -> assert false (* unparsable *)
```
This could still come from a ppx and I think we usually try to do better than an `assert false` in such cases.

Octachron replied:
True. Since this seemed to be a parsing error, I made this requirement an ast invariant. I thought about using using a hand-tailored type for the right hand side of module type declaration, but that required quite a few changes to the various parsetree mappers and iterators with didn't seem really useful. .

lpw25 commented on typing/typemod.ml:
```diff
@@ -1398,6 +1510,22 @@ and transl_signature env sg =
+              | Mty_ident _ -> `Substituted_away subst
+              | _ -> `Unpackable_modtype_substituted_away (id,subst)
+            in
+            let newenv, mtd, _sg = transl_modtype_decl ~info names env pmtd in
```
Rather than adding an `info` parameter to `transl_modtype_decl` I think it would be cleaner and more consistent to lift the call to `Signature_names.check_modtype` out of `transl_modtype_decl` to its call sites.

lpw25 reviewed (approved):
This all looks correct. A few small improvements for clarity and this will be good to go.

yallop commented:
This PR fixes the recently bot-closed #5460 (which has some discussion of use cases).

Octachron committed:
review: rename Pwith_module_type* to Pwith_modtype*

Octachron committed:
review: printing parsetree

Octachron committed:
review: error message

Octachron committed:
review: iterator extensions

Octachron committed:
review: begin...end for a then

Octachron committed:
review: add a parsetree invariant

module type substitution (module type T := MT) must have
a right hand side.

Octachron committed:
review: lift check_modtype calls

Octachron committed:
review: merge_constraints remove recursive knot

Octachron committed:
update changes

lpw25 commented on parsing/pprintast.ml:
```diff
@@ -1162,6 +1169,16 @@ and signature_item ctxt f x : unit =
+      pp f "@[<hov2>module@ type@ %s%a@]%a"
+        s.txt
+        (fun f md -> match md with
+           | None -> ()
```
Maybe assert false here

Octachron replied:
Fixed. I have also removed the anonymous function that was only here to handle this case.

lpw25 reviewed (approved):
I think this is good to merge now.
